packages/themeables/core/src/index.ts

```diff
@@ -2,44 +2,40 @@ import { createContext, useContext, createElement } from 'react'
 import type { FC } from 'react'
 import { component, startWithType } from 'refun'
 
-export type TThemeables<ThemeType, ComponentMappings> = { [key in keyof ComponentMappings]: (props: ComponentMappings[key]) => ThemeType }
-export type TOverrideables<ThemeType, ComponentMappings> = { [key in keyof ComponentMappings]?: (props: ComponentMappings[key]) => Partial<ThemeType> }
-export type TOverrideContext<ThemeType, ComponentMappings> = React.Context<TOverrideables<ThemeType, ComponentMappings>>
+export type TThemeables<ThemeType, ComponentMappings> = { [key in keyof ComponentMappings]?: (props: ComponentMappings[key]) => Partial<ThemeType> }
 
-export const setupTheme = <ThemeType, ComponentMappings>(
-  defaultTheme: TThemeables<ThemeType, ComponentMappings>,
-  overrideTheme = {} as TOverrideContext<ThemeType, ComponentMappings>
-) => {
-  const ThemePiece = createContext(defaultTheme)
+export const setupTheme = <ThemeType>(overrideTheme = {}) => {
+  const ThemePiece = createContext({})
 
-  type K = keyof ComponentMappings
-
-  const createThemeable = <P extends ThemeType>(name: K, Target: FC<any>) => {
+  const createThemeable = <P extends ThemeType, ComponentMappings>(
+    name: keyof ComponentMappings,
+    Target: FC<any>
+  ) => {
     const Themeable = component(
-      startWithType<Partial<P> & ComponentMappings[K]>(),
+      startWithType<Partial<P> & ComponentMappings[keyof ComponentMappings]>(),
       (props) => {
-        const themeProps = useContext(ThemePiece)[name](props)
-        const overrideContext = useContext(overrideTheme)
-
-        if (overrideContext && typeof overrideContext[name] === 'function') {
-          return {
-            ...themeProps,
-            ...overrideContext[name]!(props),
-            ...props,
-          }
+        let themeProps = {}
+        let overrideProps = {}
+
+        const themeContext = useContext(ThemePiece as React.Context<TThemeables<ThemeType, ComponentMappings>>)
```

Comment:
omg, i'm trying to understand that line for 15min now

Reply:
`themeContext` is the object that is stored in the `ThemePiece` context. Without TypeScript, we would only have this:
`const themeContext = useContext(ThemePiece)`
the whole part of ` as React.Context<TThemeables<ThemeType, ComponentMappings>>` assigns a special type to `ThemePiece`. So it gets a type of React.Context, which takes a "type prop" (aka generic) or TThemeables (which is the type of object stored in the context). But TThemeable is not a regular type. It's more complex and it requires it's own "type props" (aka generics) ThemeType & ComponentMappings.

We couldn't do it at the moment of creation of `ThemePiece` because inside `setupTheme()` we didn't have the `ThemeType` & `ComponentMappings` - we only have access to them inside `createThemeable()`, so that's why we're only assigning the type in this line with the use of `as`.

Reply:
If you want to we can have a quick call to talk about this line :)

Reply:
<img width="1088" alt="Screenshot 2020-10-23 at 13 54 25" src="https://user-images.githubusercontent.com/16437281/97000693-62ce3d80-1537-11eb-8cf8-8079e7d753c4.png">
😅

Reply:
> this will happen when you know the type of some entity could be more specific than it's current type
Yes, this is exactly what happens here. We needed to allow empty object first, because we don't have access to specific type and the moment we get access to specific type we just enforce it :D